### ISSUE TITLE
Gpt-neo inference with Deepspeed: IndexError: Dimension out of range

### ISSUE THREAD

**Issue by kingpalethe** [Xirider/finetune-gpt2xl#15]
Thanks for this useful repository. I was able to follow it to train a gtp-neo 2.7B model. 

Inference on the model works well for me, using less than 8GB of Vram, so fits on consumer-level gpus, however, I'm not yet able to get the inference working with Deepspeed. 

To be clear... 

I am using the code from here:

https://github.com/Xirider/finetune-gpt2xl/blob/main/README.md#generate-text-with-a-gpt-neo-27-billion-parameters-model

And it works well, if I comment out this line:

`deepspeed.init_inference(model, mp_size=1, dtype=torch.half, replace_method='auto')`

If I retain the line, then the inference fails with this error message:

```
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/ops/transformer/inference/transformer_inference.py", line 374, in forward
    output = DeepSpeedSelfAttentionFunction.apply(
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/ops/transformer/inference/transformer_inference.py", line 312, in forward
    output, key_layer, value_layer, context_layer = selfAttention_fp()
  File "/opt/conda/lib/python3.8/site-packages/deepspeed/ops/transformer/inference/transformer_inference.py", line 270, in selfAttention_fp
    qkv_out = qkv_func(input,
IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 2)
```

I'm actually a bit vague on whether Deepspeed actually should be used with inference for GTP-NEO, as far.

Huggingface says....

https://huggingface.co/transformers/main_classes/deepspeed.html

> DeepSpeed ZeRO-2 is primarily used only for training, as its features are of no use to inference.

But, Microsoft has a guide which shows the usage of Deepspeed for inference with this model...

https://github.com/microsoft/DeepSpeed/blob/master/docs/_tutorials/inference-tutorial.md#end-to-end-gpt-neo-27b-inference

**Comment by kingpalethe** [comment]
After further testing, I am finding that inference with Deepspeed DOES offer the promised ~2x speedup, and this code ...

https://github.com/Xirider/finetune-gpt2xl/blob/main/README.md#generate-text-with-a-gpt-neo-27-billion-parameters-model

DOES work, but only if I use the not-finetuned, original model, from Huggingface -- "EleutherAI/gpt-neo-2.7B"..

so:

```
model = GPTNeoForCausalLM.from_pretrained("EleutherAI/gpt-neo-2.7B").half().to("cuda")
tokenizer = AutoTokenizer.from_pretrained("EleutherAI/gpt-neo-2.7B")
```
works well.

But when I used my custom finetuned model, which I finetuned according to this guide... https://github.com/Xirider/finetune-gpt2xl/blob/main/README.md#finetune-gpt-neo-27-billion-parameters
....
I get the 
```IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 2)
```
error when I try inference.

Again, inference on my finetuned model, WITHOUT deepspeed, works well.

So in short it seems that something about the finetune process shown here produces a finetuned model that doesn't seem to work with Deepspeed inference, even through the original (not-fintuned) model can be shown to work will deepspeed inference.

**Comment by kingpalethe** [comment]
I've found what seems to be a solution, that enables Deepspeed inference on my trained gpt-neo model.

The issue seems to be that the trained gpt-neo has this line in its `config.json`:

`"use_cache": false,`

I can see that the original model -- "EleutherAI/gpt-neo-2.7B", this value is set to `false`
https://huggingface.co/EleutherAI/gpt-neo-2.7B/blob/main/config.json#L77

So it seems that if, after training, I manually modify `config.json`, in the folder of my trained model, and set this value to `true`, then Deepspeed inference works as expected. I get about a 25% speedup on inference on an rtx 2070 super.

Closing this.

**Comment by martingajek** [comment]
Thanks for the tip, it works!
